File: backend/src/Controllers/PackageController.php
<?php

declare(strict_types=1);

namespace Quantis\AIPortfolioAssistant\Controllers;

use Quantis\AIPortfolioAssistant\Services\PackageResolver;
use PDO;

/**
 * Package Controller
 *
 * Exposes the effective role-based capability package to the frontend, and
 * provides admin CRUD for editing packages via gpt_admin.
 *
 * Routes:
 *   GET  /api/v1/me/package                - current user's effective package
 *   GET  /api/v1/admin/packages            - all 4 packages (admin only)
 *   GET  /api/v1/admin/packages/{role}     - single package (admin only)
 *   PUT  /api/v1/admin/packages/{role}     - update capabilities (admin only)
 */
class PackageController
{
    private PDO $db;
    private array $config;
    private PackageResolver $resolver;

    public function __construct(PDO $db, array $config)
    {
        $this->db = $db;
        $this->config = $config;
        $this->resolver = new PackageResolver($db);
    }

    /**
     * Return the package for the currently authenticated user. Callers include
     * both gpt (to render sidebar, filter providers) and gpt_admin (to confirm
     * the caller is an admin before unlocking the UI).
     */
    public function me(array $request): array
    {
        $userId = $request['user_id'] ?? null;
        $userIdInt = is_numeric($userId) ? (int) $userId : null;

        $package = $this->resolver->resolveForUser($userIdInt);

        return [
            'success' => true,
            'role' => $package['role'],
            'capabilities' => $package['capabilities'],
            'updated_at' => $package['updated_at'],
        ];
    }

    /**
     * List all 4 packages. Admin only.
     */
    public function adminList(array $request): array
    {
        if ($err = $this->requireAdmin($request)) {
            return $err;
        }

        $packages = [];
        foreach (PackageResolver::validRoles() as $role) {
            $packages[] = $this->resolver->loadPackage($role);
        }

        return [
            'success' => true,
            'packages' => $packages,
        ];
    }

    /**
     * Get a single package by role. Admin only.
     */
    public function adminGet(array $request, string $role): array
    {
        if ($err = $this->requireAdmin($request)) {
            return $err;
        }

        if (!in_array($role, PackageResolver::validRoles(), true)) {
            return [
                'success' => false,
                'error' => 'Invalid role',
                'status_code' => 400,
            ];
        }

        return [
            'success' => true,
            'package' => $this->resolver->loadPackage($role),
        ];
    }

    /**
     * Replace the capabilities JSON for a role. Admin only.
     * Expects body: { "capabilities": { ... } }
     */
    public function adminUpdate(array $request, string $role): array
    {
        if ($err = $this->requireAdmin($request)) {
            return $err;
        }

        if (!in_array($role, PackageResolver::validRoles(), true)) {
            return [
                'success' => false,
                'error' => 'Invalid role',
                'status_code' => 400,
            ];
        }

        $body = $request['body'] ?? [];
        $capabilities = $body['capabilities'] ?? null;

        if (!is_array($capabilities)) {
            return [
                'success' => false,
                'error' => 'capabilities object is required',
                'status_code' => 400,
            ];
        }

        $validationError = $this->validateCapabilities($capabilities);
        if ($validationError !== null) {
            return [
                'success' => false,
                'error' => $validationError,
                'status_code' => 400,
            ];
        }

        $encoded = json_encode($capabilities, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        if ($encoded === false) {
            return [
                'success' => false,
                'error' => 'Failed to encode capabilities: ' . json_last_error_msg(),
                'status_code' => 400,
            ];
        }

        $adminId = is_numeric($request['user_id'] ?? null) ? (int) $request['user_id'] : null;

        // Atomic upsert via INSERT ... ON DUPLICATE KEY UPDATE.
        //
        // Why not "UPDATE first, INSERT if rowCount() === 0":
        //   MySQL's PDO::rowCount() for UPDATE returns the number of rows
        //   that were *modified*, NOT the number that *matched*. So if the
        //   admin clicks Save twice in a row (or saves with no actual
        //   capability change), the second UPDATE returns rowCount=0 even
        //   though the row exists — the fallback then runs INSERT and hits
        //   "Duplicate entry 'X' for key 'packages.PRIMARY'".
        //   This was reported as a 500 in the gpt_admin Packages editor.
        //
        // ON DUPLICATE KEY UPDATE handles both insert and update in one
        // statement and is rowCount-semantics-immune: if the row exists,
        // capabilities + updated_by get overwritten; if not, the row is
        // created. Either way, exactly one final row.
        $stmt = $this->db->prepare(
            'INSERT INTO packages (role, capabilities, updated_by) VALUES (:role, :caps, :uid)
             ON DUPLICATE KEY UPDATE capabilities = VALUES(capabilities), updated_by = VALUES(updated_by)'
        );
        $stmt->execute([
            ':role' => $role,
            ':caps' => $encoded,
            ':uid' => $adminId,
        ]);

        return [
            'success' => true,
            'package' => $this->resolver->loadPackage($role),
        ];
    }

    /**
     * Enforce that the current request is authenticated AND the user's role is admin.
     * Mirrors SystemSettingsController::requireAdmin.
     */
    private function requireAdmin(array $request): ?array
    {
        $userId = $request['user_id'] ?? null;
        if (!$userId) {
            return [
                'success' => false,
                'error' => 'Authentication required',
                'status_code' => 401,
            ];
        }

        $stmt = $this->db->prepare('SELECT role FROM users WHERE id = :id LIMIT 1');
        $stmt->execute([':id' => $userId]);
        $user = $stmt->fetch(PDO::FETCH_ASSOC);

        if (!$user || ($user['role'] ?? '') !== 'admin') {
            return [
                'success' => false,
                'error' => 'Admin access required',
                'status_code' => 403,
            ];
        }

        return null;
    }

    /**
     * Shallow shape check. Deeper validation stays loose so gpt_admin can add
     * new knobs (e.g. new provider names) without needing a backend change.
     */
    private function validateCapabilities(array $caps): ?string
    {
        foreach (['providers', 'sidebar'] as $required) {
            if (!array_key_exists($required, $caps) || !is_array($caps[$required])) {
                return "capabilities.{$required} must be an object";
            }
        }

        foreach (['mcp_servers', 'skills'] as $optional) {
            if (array_key_exists($optional, $caps)
                && $caps[$optional] !== null
                && !is_array($caps[$optional])) {
                return "capabilities.{$optional} must be null or an array";
            }
        }

        if (array_key_exists('quota_tokens', $caps)
            && $caps['quota_tokens'] !== null
            && !is_int($caps['quota_tokens'])) {
            return 'capabilities.quota_tokens must be null or an integer';
        }

        return null;
    }
}

File: backend_python/app/controllers/package_controller.py
"""Port of Controllers/PackageController.php."""
from __future__ import annotations

import json

from app.services.package_resolver import PackageResolver
from app.support.phpcompat import is_numeric


class PackageController:
    def __init__(self, db, config):
        self.db = db
        self.config = config
        self.resolver = PackageResolver(db)

    def me(self, request) -> dict:
        uid = request.get('user_id')
        package = self.resolver.resolveForUser(int(uid) if is_numeric(uid) else None)
        return {'success': True, 'role': package['role'], 'capabilities': package['capabilities'],
                'updated_at': package['updated_at']}

    def adminList(self, request) -> dict:
        err = self._require_admin(request)
        if err:
            return err
        return {'success': True, 'packages': [self.resolver.loadPackage(r) for r in PackageResolver.validRoles()]}

    def adminGet(self, request, role: str) -> dict:
        err = self._require_admin(request)
        if err:
            return err
        if role not in PackageResolver.validRoles():
            return {'success': False, 'error': 'Invalid role', 'status_code': 400}
        return {'success': True, 'package': self.resolver.loadPackage(role)}

    def adminUpdate(self, request, role: str) -> dict:
        err = self._require_admin(request)
        if err:
            return err
        if role not in PackageResolver.validRoles():
            return {'success': False, 'error': 'Invalid role', 'status_code': 400}
        caps = (request.get('body') or {}).get('capabilities')
        if not isinstance(caps, (dict, list)):
            return {'success': False, 'error': 'capabilities object is required', 'status_code': 400}
        v = self._validate_capabilities(caps)
        if v is not None:
            return {'success': False, 'error': v, 'status_code': 400}
        encoded = json.dumps(caps, ensure_ascii=False, separators=(',', ':'))
        admin_id = int(request['user_id']) if is_numeric(request.get('user_id')) else None
        self.db.execute(
            'INSERT INTO packages (role, capabilities, updated_by) VALUES (:role, :caps, :uid)'
            ' ON DUPLICATE KEY UPDATE capabilities = VALUES(capabilities), updated_by = VALUES(updated_by)',
            {':role': role, ':caps': encoded, ':uid': admin_id})
        self.resolver = PackageResolver(self.db)   # PHP re-reads because loadPackage cache is per request
        return {'success': True, 'package': self.resolver.loadPackage(role)}

    def _require_admin(self, request) -> dict | None:
        uid = request.get('user_id')
        if not uid:
            return {'success': False, 'error': 'Authentication required', 'status_code': 401}
        user = self.db.fetch_one('SELECT role FROM users WHERE id = :id LIMIT 1', {':id': uid})
        if not user or (user.get('role') or '') != 'admin':
            return {'success': False, 'error': 'Admin access required', 'status_code': 403}
        return None

    @staticmethod
    def _validate_capabilities(caps) -> str | None:
        if isinstance(caps, list):
            caps = {}   # PHP array with no keys: required keys missing → first error below
        for required in ('providers', 'sidebar'):
            if required not in caps or not isinstance(caps[required], (dict, list)):
                return f'capabilities.{required} must be an object'
        for optional in ('mcp_servers', 'skills'):
            if optional in caps and caps[optional] is not None and not isinstance(caps[optional], (dict, list)):
                return f'capabilities.{optional} must be null or an array'
        if 'quota_tokens' in caps and caps['quota_tokens'] is not None and \
                (isinstance(caps['quota_tokens'], bool) or not isinstance(caps['quota_tokens'], int)):
            return 'capabilities.quota_tokens must be null or an integer'
        return None
